modules/mqtt/destination/mqtt-destination.c

```diff
@@ -0,0 +1,281 @@
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "mqtt-destination.h"
+#include "mqtt-worker.h"
+
+#include "plugin.h"
+#include "messages.h"
+#include "misc.h"
+#include "stats/stats-registry.h"
+#include "logqueue.h"
+#include "driver.h"
+#include "plugin-types.h"
+#include "logthrdest/logthrdestdrv.h"
+
+/*
+ * Default values
+ */
+
+#define DEFAULT_ADDRESS "tcp://localhost:1883"
+#define DEFAULT_CLEAN_SESSION FALSE
+#define DEFAULT_KEEPALIVE 60
+#define DEFAULT_QOS 0
+#define DEFAULT_MESSAGE_TEMPLATE "$MESSAGE"
+
+/*
+ * Configuration
+ */
+
+void
+mqtt_dd_set_topic_template(LogDriver *d, LogTemplate *topic)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+
+  log_template_unref(self->topic_name);
+  self->topic_name = topic;
+}
+
+void
+mqtt_dd_set_fallback_topic(LogDriver *d, const gchar *fallback_topic)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+
+  g_free(self->fallback_topic_name);
+  self->fallback_topic_name = g_strdup(fallback_topic);
+}
+
+void
+mqtt_dd_set_keepalive (LogDriver *d, const gint keepalive)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  self->keepalive = keepalive;
+}
+
+void
+mqtt_dd_set_address(LogDriver *d, const gchar *address)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+
+  g_string_assign(self->address, address);
+}
+
+void
+mqtt_dd_set_qos (LogDriver *d, const gint qos)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  self->qos = qos;
+}
+
+void
+mqtt_dd_set_message_template_ref(LogDriver *d, LogTemplate *message)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+
+  log_template_unref(self->message);
+  self->message = message;
+}
+
+/*
+ * Utilities
+ */
+static const gchar *
+_format_stats_instance(LogThreadedDestDriver *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  static gchar stats_instance[1024];
+
+  if (((LogPipe *)d)->persist_name)
+    g_snprintf(stats_instance, sizeof(stats_instance), "%s", ((LogPipe *)d)->persist_name);
+  else
+    g_snprintf(stats_instance, sizeof(stats_instance), "mqtt,%s,%s", self->address->str,
+               self->topic_name->template);
+
+  return stats_instance;
+}
+
+static const gchar *
+_format_persist_name(const LogPipe *d)
+{
+  MQTTDestinationDriver *self = (MQTTDestinationDriver *)d;
+  static gchar persist_name[1024];
+
+  if (d->persist_name)
+    g_snprintf(persist_name, sizeof(persist_name), "mqtt-destination.%s", d->persist_name);
+  else
+    g_snprintf(persist_name, sizeof(persist_name), "mqtt-destination.(%s,%s)", self->address->str,
+               self->topic_name->template);
+
+  return persist_name;
+}
+
+static void
+_set_default_value(MQTTDestinationDriver *self, GlobalConfig *cfg)
+{
+  self->address       = g_string_new(DEFAULT_ADDRESS);
```

Comment:
This leaks memory. I've found this with valgrind.

It's worth organizing your commits by small functionalities you add step by step.
This way, the missing `g_string_free()` could have been seen trivially when reviewing the commit that adds it.